src/chains/ethereum/ethereum/src/block-time.ts

```diff
@@ -0,0 +1,100 @@
+   */
+  getOffset(): number;
+
+  /**
+   * Set the current offset of the BlockTime instance. This is the value that would be added to the system time in order to result in the desired timestamp.
+   * @param {number} offset - The "offset" for the BlockTime instance in milliseconds.
+   */
+  setOffset(offset: number): void;
+
+  /**
+   * Create a timestamp for a new block in seconds. Can specify the timestamp for the block, in milliseconds, which will be reflected in future calls to this function.
+   * @param  {number} timestamp? - Optional - The timestamp for the block that will be mined, in milliseconds.
+   * @returns number - the timestamp in seconds
+   */
+  createBlockTimestampInSeconds(timestamp?: number): number;
+}
+
+export class ClockBasedBlockTime implements BlockTime {
+  private _getReferenceClockTime: () => number;
+  private _timeOffset: number = 0;
+
+  private static validateTimestamp(timestamp: number) {
+    if (timestamp < 0) {
+      throw new Error(`Invalid timestamp: ${timestamp}. Value must be positive.`);
+    }
+  }
+
+  constructor(getReferenceClockTime: () => number, startTime: number | undefined) {
+    this._getReferenceClockTime = getReferenceClockTime;
+
+    if (startTime !== undefined) {
+      ClockBasedBlockTime.validateTimestamp(startTime);
+      this.setTime(startTime);
+    }
+  }
+
+  getOffset(): number {
+    return this._timeOffset;
+  }
+
+  setOffset(offset: number) {
+    const referenceTimestamp = this._getReferenceClockTime();
+    if (offset < -referenceTimestamp) {
+      throw new Error(`Invalid offset: ${offset}, value must be greater than the negative of the current reference clock timestamp: ${referenceTimestamp}`);
+    }
+    this._timeOffset = offset;
+  }
+
+  setTime(timestamp: number): number {
+    ClockBasedBlockTime.validateTimestamp(timestamp);
+    this._timeOffset = timestamp - this._getReferenceClockTime();
+    return this._timeOffset;
+  }
+
+  createBlockTimestampInSeconds(timestamp?: number): number {
+    let milliseconds;
+    if (timestamp != undefined) {
+      this.setTime(timestamp);
+      milliseconds = timestamp;
+    } else {
+      milliseconds = this._getReferenceClockTime() + this._timeOffset;
+    }
+
+    const seconds = Math.floor(milliseconds / 1000);
+    return seconds;
+  }
+}
+
+/*
+  A BlockTime implementation that increments it's reference time by the duration specified by incrementMilliseconds,
+  every time createBlockTimestampInSeconds() is called. The timestamp returned will be impacted by the timestamp offset,
+  so can be moved forward and backwards by calling .setOffset() independent of the start time, and will increment when
+  createBlockTimestampInSeconds() is called.
+*/
+export class IncrementBasedBlockTime extends ClockBasedBlockTime {
+  private readonly _tickReferenceClock: () => void;
+
+  constructor(startTime: number, incrementMilliseconds: number) {
+    let referenceTime = startTime;
```

Comment:
I don't think this `referenceTime` is reset on `evm_revert`, but I think it should be.

Reply:
That's a really good point. I rethought the approach and went back and forward a couple of times. I've settled on a slight modification, which I think makes sense. The IncrementBasedBlockTime now holds the reference clock at `startTime` and increments it's _offset_ rather than reference clock.

Offset wasn't working correctly with increment based clocks prior to this change, (as `blocktime = previousBlocktime + increment + offset` will result in compounding the offset). This change fixes that.

It does mean that the meaning of "offset" wrt increment based clocks is now strictly the elapsed duration (in block time) since the "start time" (plus / minus any modifications made by calls to `evm_setTime` / `evm_increaseTime`).

Happy to hear opinions about the above, or we can set up a call to beat out the details.